connectivity/check/context.go

```diff
@@ -0,0 +1,479 @@
+	total := ct.tests
+	actions := ct.actions()
+	skipped := ct.skippedTests()
+	failed := ct.failedTests()
+
+	nt := len(total)
+	na := len(actions)
+	ns := len(skipped)
+	nf := len(failed)
+	nw := ct.warnings()
+
+	if nf > 0 {
+		ct.Header("📋 Test Report")
+
+		// There are failed tests, fetch all failed actions.
+		fa := len(ct.failedActions())
+
+		ct.Failf("%d/%d tests failed (%d/%d actions), %d warnings, %d tests skipped:", nf, nt-ns, fa, na, nw, ns)
+
+		// List all failed actions by test.
+		for _, t := range failed {
+			ct.Logf("Test [%s]:", t.Name())
+			for _, a := range t.failedActions() {
+				ct.Log("  ❌", a)
+			}
+		}
+
+		return fmt.Errorf("%d tests failed", nf)
+	}
+
+	ct.Headerf("✅ All %d tests (%d actions) successful, %d warnings, %d tests skipped.", nt-ns, na, nw, ns)
+
+	return nil
+}
+
+func (ct *ConnectivityTest) enableHubbleClient(ctx context.Context) error {
+	ct.Log("🔭 Enabling Hubble telescope...")
+
+	dialCtx, cancel := context.WithTimeout(ctx, 2*time.Second)
+	defer cancel()
+
+	c, err := grpc.DialContext(dialCtx, ct.params.HubbleServer, grpc.WithInsecure())
+	if err != nil {
+		return err
+	}
+
+	ct.hubbleClient = observer.NewObserverClient(c)
+
+	status, err := ct.hubbleClient.ServerStatus(ctx, &observer.ServerStatusRequest{})
+	if err != nil {
+		ct.Warn("Unable to contact Hubble Relay, disabling Hubble telescope and flow validation:", err)
+		ct.Info("Expose Relay locally with: kubectl port-forward -n kube-system deployment/hubble-relay 4245:4245")
+		ct.hubbleClient = nil
+		ct.params.Hubble = false
+
+		if ct.params.FlowValidation == FlowValidationModeStrict {
+			ct.Fail("In --flow-validation=strict mode, Hubble must be available to validate flows")
+			return fmt.Errorf("hubble is not available: %w", err)
+		}
+	} else {
+		ct.Infof("Hubble is OK, flows: %d/%d", status.NumFlows, status.MaxFlows)
+	}
+
+	return nil
+}
+
+func (ct *ConnectivityTest) logAggregationMode(ctx context.Context, client *k8s.Client) (string, error) {
+	cm, err := client.GetConfigMap(ctx, ct.params.CiliumNamespace, defaults.ConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("unable to retrieve ConfigMap %q: %w", defaults.ConfigMapName, err)
+	}
+
+	if cm.Data == nil {
+		return "", fmt.Errorf("ConfigMap %q does not contain any configuration", defaults.ConfigMapName)
+	}
+
+	return cm.Data[defaults.ConfigMapKeyMonitorAggregation], nil
+}
+
+// initClients checks if Cilium is installed on the cluster, whether the cluster
+// has multiple nodes, and whether or not monitor aggregation is enabled.
+// TODO(timo): Split this up, it does a lot.
+func (ct *ConnectivityTest) initClients(ctx context.Context) error {
+	c := &deploymentClients{
+		src: ct.client,
+		dst: ct.client,
+	}
+
+	if a, _ := ct.logAggregationMode(ctx, c.src); a != defaults.ConfigMapValueMonitorAggregatonNone {
+		ct.flowAggregation = true
+	}
+
+	if ct.params.MultiCluster != "" && ct.params.SingleNode {
```

Comment:
I was thinking that multi-cluster mode would benefit from a better abstractions, something without so many ifs.

Reply:
This line is literally for excluding incompatible command line arguments. From hereon it should suffice to check only 'ct.params.MultiCluster'.